adage/analyze/models.py

```diff
@@ -0,0 +1,252 @@
+        "sample name",
+        max_length=80,
+        blank=False)
+    ml_data_source = models.CharField(
+        "Machine Learning data used for modeling, e.g. CEL file",
+        max_length=120,
+        null=True,
+        unique=True,
+        blank=True)
+    experiments = models.ManyToManyField(Experiment)
+
+    def __str__(self):
+        return "%d (%s)" % (self.id, self.name)
+
+    def get_annotation_dict(self):
+        return SampleAnnotation.objects.get_as_dict(self)
+
+    def get_annotation_items(self):
+        return iter(SampleAnnotation.objects.get_as_dict(self).items())
+
+
+class AnnotationTypeManager(models.Manager):
+    def create(self, typename, description=''):
+        ann_type = AnnotationType(
+            typename=typename,
+            description=description,
+        )
+        ann_type.full_clean()
+        ann_type.save()
+        return ann_type
+
+    def get_or_create(self, typename, description=''):
+        try:
+            ann_type = self.get(typename=typename)
+            created = False
+        except ObjectDoesNotExist:
+            ann_type = self.create(
+                typename=typename,
+                description=description)
+            created = True
+        return (ann_type, created)
+
+
+class AnnotationType(models.Model):
+    typename = models.CharField(
+        "name for this AnnotationType (usable as a Python identifier)",
+        max_length=40,
+        unique=True,
+        blank=False,
+        validators=[validate_pyname])
+    description = models.CharField(
+        max_length=140,
+        blank=True)
+
+    objects = AnnotationTypeManager()
+
+    def __str__(self):
+        return "%d (%s)" % (self.id, self.typename)
+
+
+class SampleAnnotationManager(models.Manager):
+    def create_from_dict(self, sample, ann_dict):
+        for k, v in ann_dict.items():
+            if not v:
+                continue
+            ann_type, created = AnnotationType.objects.get_or_create(k)
+            sa = SampleAnnotation(
+                sample=sample,
+                annotation_type=ann_type,
+                text=v,
+            )
+            sa.save()
+
+    def get_as_dict(self, sample):
+        annotations_for_sample = self.get_queryset().filter(sample=sample)
+        result = {sa.annotation_type.typename: sa.text
+                  for sa in annotations_for_sample}
+        return result
+
+
+class SampleAnnotation(models.Model):
+    annotation_type = models.ForeignKey(
+        AnnotationType,
+        on_delete=models.PROTECT)
+    sample = models.ForeignKey(
+        Sample,
+        on_delete=models.CASCADE)
+    text = models.TextField(
+        "annotation text",
+        blank=True)
+
+    objects = SampleAnnotationManager()
+
+    class Meta:
+        unique_together = (("annotation_type", "sample"),)
+
+    def __str__(self):
+        return "%d: (%s for %s)" % \
+            (self.id, self.annotation_type.typename, self.sample.name)
```

Comment:
Line 134 has a continuation character that escaped our previous purges. I think this can be written with parentheses to span the two lines like so:

```suggestion
        return ("%d: (%s for %s)" % 
            (self.id, self.annotation_type.typename, self.sample.name))
```